espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/dispatcher/KeyDispatch.kt

```diff
@@ -0,0 +1,160 @@
+            // 12: Append action object with subtype property changed to 'keyUp' to cancel list
+            val cancelActionObject = ActionObject(actionObject)
+            cancelActionObject.subType = InputSource.ActionType.KEY_UP
+            inputStateTable.addActionToCancel(cancelActionObject)
+
+            // 13: Call implementation specific key-down event
+            dispatcherAdapter.keyDown(keyEvent)
+        } else {
+            // 13: Call implementation specific key-up event
+            dispatcherAdapter.keyUp(keyEvent)
+        }
+    } finally {
+        dispatcherAdapter.unlockAdapter()
+    }
+
+    // Log the updated keyboard state
+    dispatcherAdapter.logger.info(String.format(
+            "Current state of key input source with id: %s",
+            inputState.logMessage()
+    ))
+    return keyEvent
+}
+
+/**
+ * Dispatch a key down event (in accordance with https://www.w3.org/TR/webdriver/#keyboard-actions)
+ *
+ * @param dispatcherAdapter Adapter that has implementation specific keyDown event
+ * @param actionObject W3C Action Object
+ * @param inputState State of the input source
+ * @param inputStateTable All of the input states for this session
+ * @param tickDuration How long the 'tick' is. This is unused currently but may be used in the future
+ * @return W3CKeyEvent (for testing purposes). 'null' if the dispatch failed.
+ */
+@Throws(AppiumException::class)
+fun dispatchKeyDown(dispatcherAdapter: W3CActionAdapter,
+                    actionObject: ActionObject, inputState: KeyInputState,
+                    inputStateTable: InputStateTable, tickDuration: Float = 0f): W3CKeyEvent? {
+    return dispatchKeyEvent(dispatcherAdapter, actionObject, inputState, inputStateTable, tickDuration, true)
+}
+
+/**
+ * Dispatch a key up event (in accordance with https://www.w3.org/TR/webdriver/#keyboard-actions)
+ *
+ * @param dispatcherAdapter Adapter that has implementation specific keyDown event
+ * @param actionObject W3C Action Object
+ * @param inputState State of the input source
+ * @param inputStateTable All of the input states for this session
+ * @param tickDuration How long the 'tick' is. This is unused currently but may be used in the future
+ * @return W3CKeyEvent (for testing purposes). 'null' if the dispatch failed.
+ */
+@Throws(AppiumException::class)
+fun dispatchKeyUp(dispatcherAdapter: W3CActionAdapter,
+                  actionObject: ActionObject, inputState: KeyInputState,
+                  inputStateTable: InputStateTable, tickDuration: Float = 0f): W3CKeyEvent? {
+    return dispatchKeyEvent(dispatcherAdapter, actionObject, inputState, inputStateTable, tickDuration, false)
+}
+
+/**
+ * Helper method to get a Key Event that has common attributes between KeyUp and KeyDispatch
+ * @param rawKey The value of the key stroke
+ * @return W3CKeyEvent Key event info used by adapter
+ */
+@Throws(AppiumException::class)
+fun getKeyEvent(dispatcherAdapter: W3CActionAdapter, rawKey: String?): W3CKeyEvent {
```

Comment:
Can write like this?
```kotlin
fun getKeyEvent(dispatcherAdapter: W3CActionAdapter, rawKey: String?): W3CKeyEvent {
    return W3CKeyEvent().apply {
        // 1. Let raw key be action's value property
        // 2. Let key be equal to the normalised key value for raw key
        this.key = instance!!.getNormalizedKey(rawKey!!)

        // 4. Let code be 'code' for raw keypress
        this.code =  getCode(rawKey)

        // 5. Let location be the key location for raw key
        this.location = getLocation(rawKey)

        // 6. Let charCode, keyCode and which be the implementation-specific values of the charCode,
        // keyCode and which properties appropriate for a key with key key and location location on
        // a 102 key US keyboard
        this.keyCode = dispatcherAdapter.getKeyCode(this.key, this.location)
        this.charCode = dispatcherAdapter.getCharCode(this.code, this.location)
        this.which = dispatcherAdapter.getWhich(this.key, this.location)
    }
```

Reply:
done, I did not know about `apply`

Reply:
👍 
It is helpful to get the class itself (as `return this`) :)